File: typescript/agentic-runtime/packages/agent-runtime/src/provider.ts
/**
 * Source: agentic-native-stack.md
 * Context: 4.8 @agentic/agent-runtime
 * Extraction ID: CODE-060
 * Knowledge Links: KI-061
 * Status: scaffolded
 */

export interface Provider {
  complete(request: ModelRequest): Promise<ModelResponse>;
}

export interface ModelRequest {
  model: string;
  messages: ModelMessage[];
  tools: string[];
  context?: Record<string, unknown>;
}

export interface ModelMessage {
  role: "system" | "user" | "assistant" | "tool";
  content: string;
}

export interface ModelResponse {
  content?: string;
  toolCall?: {
    name: string;
    args: unknown;
  };
}

File: rust/agentic-core/crates/agentic-agent-core/src/provider.rs
//! Source: agentic-native-stack.md
//! Context: 3.10 agentic-agent-core
//! Extraction ID: CODE-032
//! Knowledge Links: KI-033
//! Status: scaffolded

#[async_trait]
pub trait Provider: Send + Sync {
    async fn complete(
        &self,
        request: ModelRequest,
    ) -> Result<ModelResponse, ProviderError>;

    fn supports_streaming(&self) -> bool {
        false
    }
}

#[derive(Debug, Clone)]
pub struct ModelRequest {
    pub model: String,
    pub messages: Vec<ModelMessage>,
    pub tools: Vec<ModelToolSpec>,
}

#[derive(Debug, Clone)]
pub struct ModelMessage {
    pub role: Role,
    pub content: String,
}

#[derive(Debug, Clone, Copy)]
pub enum Role {
    System,
    User,
    Assistant,
    Tool,
}

#[derive(Debug, Clone)]
pub struct ModelToolSpec {
    pub name: String,
}

#[derive(Debug, Clone)]
pub struct ModelResponse {
    pub messages: Vec<ModelMessage>,
    pub tool_call: Option<ToolCall>,
}

#[derive(Debug, Clone)]
pub struct ToolCall {
    pub name: String,
    pub args: Value,
}